File: Store/src/Sales.Domain/Model/Discounts/Discount.cs
using System;
using System.Runtime.CompilerServices;
using Sales.Domain.Model.Orders;

namespace Sales.Domain.Model.Discounts
{
    public class Discount
    {
        public long Id { get; private set; }
        public DateTime ExpirationTime { get; private set; }
        public DiscountStrategy Strategy { get; private set; }
        public long? MaxDiscountValue { get; private set; }
        public Discount(long id, DateTime expirationTime, DiscountStrategy strategy, long? maxDiscountValue)
        {
            Id = id;
            ExpirationTime = expirationTime;
            Strategy = strategy;
            MaxDiscountValue = maxDiscountValue;
        }
        public long CalculateDiscountFor(Order order)
        {
            var discount = Strategy.CalculateDiscount(order.TotalPrice());
            if (DiscountExceedsMaximumValue(discount))
                return this.MaxDiscountValue.Value;
            return discount;
        }

        private bool DiscountExceedsMaximumValue(long discount)
        {
            if (!MaxDiscountValue.HasValue) return false;
            return MaxDiscountValue.Value < discount;
        }
    }
}

File: src/main/java/store/sales/domain/model/discounts/Discount.java
package store.sales.domain.model.discounts;

import store.sales.domain.model.orders.Order;

import java.time.LocalDateTime;

public class Discount {
    private final long id;
    private final LocalDateTime expirationTime;
    private final DiscountStrategy strategy;
    private final int maxDiscountValue;

    public Discount(
            long id, LocalDateTime expirationTime, DiscountStrategy strategy, int maxDiscountValue) {
        this.id = id;
        this.expirationTime = expirationTime;
        this.strategy = strategy;
        this.maxDiscountValue = maxDiscountValue;
    }

    public int calculateDiscountFor(Order order) {
        var discount = strategy.calculateDiscount(order.totalPrice());
        if (discountExceedsMaximumValue(discount))
            return this.maxDiscountValue;
        return discount;
    }

    private boolean discountExceedsMaximumValue(int discount) {
        if (maxDiscountValue == 0) return false;
        return maxDiscountValue < discount;
    }

    public long getId() {
        return id;
    }

    public LocalDateTime getExpirationTime() {
        return expirationTime;
    }

    public DiscountStrategy getStrategy() {
        return strategy;
    }

    public int getMaxDiscountValue() {
        return maxDiscountValue;
    }
}
